### PR TITLE
Cross Chain Token List - #3pc5uq2

### PR DESCRIPTION
**Pull request by gitChimp88**
Branched from expandable table row
Adding function to use solidity multi-call contract
fetch users balances
arrange data and show in table expandable row

### CODE REVIEW

**Comment by nunocgoncalves** [comment]
Task linked: [CU-3pc5uq2 Cross Chain Token List](https://app.clickup.com/t/3pc5uq2)

File: config-overrides.js [hunk 1]
```diff
@@ -18,5 +18,18 @@ module.exports = function override(config) {
...
+  config.module.rules.push({
+    test: /\.(js|mjs|jsx)$/,
+    enforce: "pre",
+    loader: require.resolve("source-map-loader"),
+    resolve: {
+      fullySpecified: false,
+    },
+  });
+
+  // Note: this is a temporary patch until 5.0.1 is released.
```
Comment (nunocgoncalves) [inline config-overrides.js:32]: Please create a task for this patch for when the webpack update comes out

File: src/components/ERC20Balance/balanceMethods/getBalances.js [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+  const { chainId } = tokens[0];
+  tokens.forEach((token) => {
+    if (token.chainId !== chainId) {
+      // eslint-disable-next-line no-console
+      console.warn(`Requested tokens have to be on same chain.`);
+      return [];
+    }
+  });
+
+  const multicallAddress = process.env.REACT_APP_MULTICALL_ADDRESS;
```
Comment (nunocgoncalves) [inline src/components/ERC20Balance/balanceMethods/getBalances.js:48]: How would we set this env for many different contract addresses? Wouldn't it be better if it was a config?

File: src/helpers/getChainDetails.ts [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+  1: {
+    rpc: process.env.REACT_APP_RPC_PROVIDER_ETH,
+    name: "Ethereum",
+    logoUri:
+      "https://storageapi.fleek.co/d1921602-c1d0-4d59-82e9-e36a2947b855-bucket/App/static/token-icons/weth.png",
+  },
+  137: {
+    rpc: process.env.REACT_APP_RPC_PROVIDER_POLYGON,
+    name: "Polygon",
+    logoUri: "https://wallet-asset.matic.network/img/tokens/matic.svg",
```
Comment (nunocgoncalves) [inline src/helpers/getChainDetails.ts:25]: We probably want to store all of this on IPFS ourselves.

File: src/hooks/useGetTokenListToQuery.tsx [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+    //   setTokenList({
+    //     ethereum: res?.data.tokens,
+    //     polygon: resPolygon?.data.tokens,
+    //   });
+    // };
+
+    // fetchTokens();
+
+    const fetchTokensDemo = async () => {
+      // token source could be aggregated lists or maybe goplus?
```
Comment (nunocgoncalves) [inline src/hooks/useGetTokenListToQuery.tsx:52]: I think we're probably going for Go+
It's not scalable to run archival nodes for all the chains.

**Review by nunocgoncalves — COMMENTED** [review]
I did a very high-level review of your code and the logic seems just fine. I'll give you more detailed feedback on some optimization we can do once I do a more detailed review.

It looks like we can benefit a lot from using types in a couple of scenarios so maybe let's start using typescript as a standard.

File: src/helpers/getChainDetails.ts (hunk 3 above)
Comment (gitChimp88) [reply src/helpers/getChainDetails.ts:25]: Yeah I agree for the logoURI's

**Review by nunocgoncalves — CHANGES_REQUESTED** [review]
- [ ] **Fix** module not found error. (Check error bellow)
image

- [ ] Refactor line 30 in "NativeBalance.tsx" to match styling rules

- [ ] Rename "NativeBalance" component to TotalBalance <br>


Notes:

- In order for this feature to be fully complete, we have to create event listeners so that the token list updates when transactions are made.

- Creating a global listener so that the token list updates even when external transactions are made is something which we also will probably develop in the future.

**Review by nunocgoncalves — COMMENTED** [review]
Looks good. Thanks for being so fast!

File: src/components/ERC20Balance/balanceMethods/checkVariants.ts [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+    // the current way will only group if they have the native token as well.
+    const userHasNativeToken: IToken[] = tokens.filter((val: any) => {
+      return val.symbol === nativeSymbol[0];
+    });
+
+    if (userHasNativeToken.length > 0) {
+      // user has native token so group with this if the price is similar
+      const currentTokenPrice = Number(token.priceUSD);
+      const nativeTokenPrice = Number(userHasNativeToken[0].priceUSD);
+      // check price is within 5 dollars either side
```
[the remark this answers is not in the document]
Comment (nunocgoncalves) [reply src/components/ERC20Balance/balanceMethods/checkVariants.ts]: > 

Will have to do some QA to evaluate this.

[the remark this answers is not in the document]
Comment (nunocgoncalves) [reply src/components/ERC20Balance/balanceMethods/checkVariants.ts]: > What do you think is a fair amount?

Will have to do some QA to check this

**Review by nunocgoncalves — APPROVED** [review]

**Comment by nunocgoncalves** [comment]
Merged to QA!

**Pull request closed without merging**